plugins/synaptics-vmm9/fu-synaptics-vmm9-device.c

```diff
@@ -0,0 +1,709 @@
+				      error)) {
+		g_prefix_error(error, "failed to send packet: ");
+		return FALSE;
+	}
+
+	/* disregard */
+	if (flags & FU_SYNAPTICS_VMM9_COMMAND_FLAG_NO_REPLY)
+		return TRUE;
+
+	/* poll for success */
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_synaptics_vmm9_device_command_cb,
+				  FU_SYNAPTICS_VMM9_DEVICE_TIMEOUT / FU_SYNAPTICS_VMM9_BUSY_POLL,
+				  FU_SYNAPTICS_VMM9_BUSY_POLL, /* ms */
+				  &helper,
+				  error)) {
+		g_prefix_error(error, "failed to wait for !busy: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_vmm9_device_setup(FuDevice *device, GError **error)
+{
+	FuSynapticsVmm9Device *self = FU_SYNAPTICS_VMM9_DEVICE(device);
+	guint32 customer_id;
+	guint32 board_id;
+	guint8 buf[4] = {0x0};
+
+#if 1
+	//	guint32 temperature;
+	g_warning("getting CHIP TEMP");
+	if (!fu_synaptics_vmm9_device_command(self,
+					      FU_SYNAPTICS_VMM9_RC_CTRL_GET_CHIP_CORE_TEMPERATURE,
+					      0x0,
+					      NULL,
+					      4,
+					      buf,
+					      4,
+					      FU_SYNAPTICS_VMM9_COMMAND_FLAG_FULL_BUFFER,
+					      error))
+		return FALSE;
+		// g_warning("SUCCESS!!!: %s", fu_device_to_string(device));
+#endif
+
+	/* read customer ID */
+	g_warning("getting customer ID");
+	if (!fu_synaptics_vmm9_device_command(self,
+					      FU_SYNAPTICS_VMM9_RC_CTRL_READ_FROM_MEMORY,
+					      FU_SYNAPTICS_VMM9_MEM_OFFSET_CUSTOMER_ID,
+					      NULL,
+					      sizeof(buf),
+					      buf,
+					      sizeof(buf),
+					      FU_SYNAPTICS_VMM9_COMMAND_FLAG_FULL_BUFFER,
+					      error))
+		return FALSE;
+	customer_id = fu_memread_uint32(buf, G_BIG_ENDIAN);
+	fu_device_add_instance_u32(device, "CID", customer_id);
+	fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "CID", NULL);
+
+	/* read board ID */
+	g_warning("getting board ID");
+	if (!fu_synaptics_vmm9_device_command(self,
+					      FU_SYNAPTICS_VMM9_RC_CTRL_READ_FROM_MEMORY,
+					      FU_SYNAPTICS_VMM9_MEM_OFFSET_BOARD_ID,
+					      NULL,
+					      sizeof(buf),
+					      buf,
+					      sizeof(buf),
+					      FU_SYNAPTICS_VMM9_COMMAND_FLAG_FULL_BUFFER,
+					      error))
+		return FALSE;
+	board_id = fu_memread_uint32(buf, G_BIG_ENDIAN);
+	fu_device_add_instance_u32(device, "BID", board_id);
+	fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "BID", NULL);
+	//	g_warning("SUCCESS!!!: %s", fu_device_to_string(device));
+
+#if 0
+{
+	guint8 buf[13] = {0};
+	if (!fu_synaptics_vmm9_device_command(self,
+					      FU_SYNAPTICS_VMM9_RC_CTRL_READ_FROM_TX_DPCD,
+					      0x0, //offset
+					      NULL,
+					      0,
+					      buf,
+					      sizeof(buf),
+					      FU_SYNAPTICS_VMM9_COMMAND_FLAG_FULL_BUFFER,
+					      error))
+		return FALSE;
+	g_warning("SUCCESS!!!: %s", fu_device_to_string(device));
+}
+#endif
+
+	//	g_warning("SUCCESS!: %s", fu_device_to_string(device));
+
+	/* TODO: get the version and other properties */
+	fu_device_set_version(device, "1.2.3");
```

Comment:
Pretty big TODO

Reply:
Yes, indeed. I did hope to read back the entire DPCD but that doesn't seem to work.

Reply:
Tracing VmmHidTool FTW. I think I've worked it out, I'll ask Synaptics if the register assumptions are valid.